### ISSUE TITLE
Nested matches of exclude_lines patterns leads to unmatched lines being excluded

### ISSUE THREAD

**Issue by ringohoffman** [coveragepy/coveragepy#1779]
**Describe the bug**

When multiple patterns of `exclude_lines` overlap (like an `abstractmethod` inside a `Protocol`, which you see below), the exclusion for the `abstractmethod` unexpectedly extends to the next coverable line, leading to an incorrect coverage report.

**To Reproduce**

`.coveragerc`:
```ini
[report]
exclude_lines =
    @(abc\.)?abstractmethod
    class .*?\(.*?Protocol.*?:
```

`a.py`:
```python
import abc
from typing import Protocol


class MyProtocol(Protocol):
    @abc.abstractmethod
    def my_method(self) -> int:
        """Discombombulate the frobnicator."""


def function() -> int:
    x = 1
    y = 2
    return x + y

```

`tests/test_a.py`:
```python
import unittest

import a

class TestA(unittest.TestCase):
    ...

```

<img width="646" alt="Screenshot 2024-05-09 at 10 57 56" src="https://github.com/nedbat/coveragepy/assets/27844407/d0148391-be3b-4ca9-a00e-af883909689d">

How can we reproduce the problem?  Please *be specific*. Don't link to a failing CI job. Answer the questions below:
1. What version of Python are you using? 
3.10
1. What version of coverage.py shows the problem? The output of `coverage debug sys` is helpful.
`coverage==7.5.1`
1. What versions of what packages do you have installed? The output of `pip freeze` is helpful.
...
1. What code shows the problem?  Give us a specific commit of a specific repo that we can check out. If you've already worked around the problem, please provide a commit before that fix.
...
1. What commands should we run to reproduce the problem? *Be specific*. Include everything, even `git clone`, `pip install`, and so on. Explain like we're five!

```console
$ python -m pytest tests/test_a.py --cov=a.py
$ coverage html
$ python -m http.server
```

**Expected behavior**
I don't expect the exclusion to extend to the next line. 

If you change `Protocol` to `abc.ABC`, it works as expected. If you remove `@(abc\.)?abstractmethod` from `exclude_lines` or modify it in a way s.t. it does not match the code, it works as expected.

**Comment by nedbat** [comment]
I see what you mean... :(

**Comment by nedbat** [comment]
Fixed in commit 96bd93071f9e6ce5e7fd5e5c41f234d302c8611b

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.5.2](https://pypi.org/project/coverage/7.5.2).
